### PR TITLE
Gpuadvsub

### LINKED ISSUE #3166: Convert AdvancedSubtensor to GpuAdvancedSubtensor1 on the GPU
@lamblin you told me you know how to make this. Can you describe it here so that someone else have a change to make it?

### PR DESCRIPTION
**Pull request by abergeron**
Finish the work in #4396 

Fixes #3166

### CODE REVIEW

File: theano/gpuarray/opt.py [hunk 1]
```diff
@@ -957,6 +958,12 @@ def local_gpua_advanced_subtensor(op, context_name, inputs, outputs):
 
 
 @register_opt('fast_compile')
+@op_lifter([tensor.AdvancedSubtensor])
+def local_gpua_advanced_subtensor_(node, context_name):
```
Comment (lamblin) [inline theano/gpuarray/opt.py]: We should update that one to work with the new GraphToGpu optimizer:
```python
@register_opt2([tensor.AdvancedSubtensor], 'fast_compile')
def local_gpua_advanced_subtensor(op, context_name, inputs, outputs):
```

File: theano/gpuarray/tests/test_subtensor.py [hunk 2]
```diff
@@ -85,3 +86,42 @@ def test_incsub_f16():
...
+class G_advancedsubtensor(test_subtensor.TestAdvancedSubtensor):
+    def shortDescription(self):
+        return None
+
+    def __init__(self, name):
+        test_subtensor.TestAdvancedSubtensor.__init__(
+            self, name,
+            shared=gpuarray_shared_constructor,
+            sub=GpuAdvancedSubtensor,
+            inc_sub=GpuIncSubtensor,
```
Comment (lamblin) [inline theano/gpuarray/tests/test_subtensor.py]: I think we should add `dimshuffle=GpuDimShuffle` here as well.

**Comment by lamblin** [comment]
Apart from those comments, which are there because the master branch changed in the mean time, it looks good to me.
Thanks!

File: theano/gpuarray/tests/test_subtensor.py (hunk 2 above)
Comment (abergeron) [inline theano/gpuarray/tests/test_subtensor.py]: The superclass doesn't have a dimshuffle parameter.  I also noticed that inc_sub should be AdvancedIncSubtensor so I removed it.

Comment (lamblin) [inline theano/gpuarray/tests/test_subtensor.py]: OK. As long as the tests pass.

**Comment by lamblin** [comment]
There is still one failing test.

**Comment by abergeron** [comment]
Yet another rebase, and the failing test is fixed.

**Comment by abergeron** [comment]
These test cases are like whack-a-mole.

**Comment by abergeron** [comment]
Now this should be good and all the tests pass.

File: theano/gpuarray/subtensor.py [hunk 3]
```diff
@@ -519,16 +519,24 @@ def perform(self, node, inputs, out_):
...
+                        # indexing.  If ap is bigger than 0 it means we
+                        # have encountered at least one array.
+                        if ap >= 0:
+                            ap -= 1
+                        # If this index is before the first array then
+                        # we will not move the array back to its
+                        # position.  Mark this by faking that there
+                        # are more than two arrays.  This is crazy
+                        # numpy behaviour so blame them.
+                        if narrays == 0:
```
Comment (lamblin) [inline theano/gpuarray/subtensor.py]: Why that new `if`? Could `narrays` have changed since the beginning of the block?

Comment (abergeron) [inline theano/gpuarray/subtensor.py]: That is the result of two compounded changes, I'll fix it.

**Pull request merged**